### ISSUE TITLE
RFC: Introduce verbose exceptions and --error-explain

### ISSUE THREAD

**Issue by luislavena** [crystal-lang/crystal#3449]
# Background

As Crystal userbase increases, it also increases the number of questions related to design decisions. Examples of those range from *why ivar cannot be inferred from call to methods on initialize* or *ivar signature vs instances of inherited types* (aka: variance and covariance)

While documentation reflects most of these points, the exceptions messages might be a bit confusing and will trigger developers to ask the same question *over* and *over* (mea culpa, have done that in the past several times)

# Idea

Introduce a verbose/explanatory field for some exceptions, so it can be used to describe in length the issue and possible solutions.

Show abbreviated information by default and offer `--error-explain` for developers to obtain the full messages, similar to current `--error-trace`.

This will provide developers an inline response, curated by Crystal creators using the proper terminology that might avoid missing existing issues when searching here or the mailing list.

# References

- Rust extended error messages [1](https://blog.rust-lang.org/2016/08/10/Shape-of-errors-to-come.html), [2](https://blog.rust-lang.org/2016/09/29/Rust-1.12.html)
- Elm compiler errors: [3](http://elm-lang.org/blog/compiler-errors-for-humans)

---

Sending this for your consideration.

Thank you :heart: :heart: :heart:

**Comment by asterite** [comment]
Sure, we can do this :-)

**Comment by rdp** [comment]
This would be nice.  If an exception is thrown "all the way out the console" I'd suggest including the verbose output in the screen output :)

**Comment by asterite** [comment]
Of course, to do this we would first need to identify current errors which aren't easily understood, and provide a better message to understand them. You can submit proposals here.

**Comment by luislavena** [comment]
Thank you @asterite, here are a few examples:

## Instance variable not initialized on any `initialize` (due method call)

```crystal
class Counter
  @count : Int32

  def initialize
    reset
  end

  def reset
    @count = 0
  end
end
```
```
Error in ./3449-1.cr:2: instance variable '@count' of Counter was not initialized in all of the 'initialize' methods, rendering it nilable

  @count : Int32
  ^~~~~~
```

## Instance variable initialized somewhere but not all `initialize` methods

```crystal
class Person
  @name : String
  @age : Int32

  def initialize(@name)
  end

  def initialize(@name, @age)
  end
end
```
```
Error in ./3449-2.cr:5: this 'initialize' doesn't explicitly initialize instance variable '@age' of Person, rendering it nilable

The instance variable '@age' is initialized in other 'initialize' methods,
and by not initializing it here it's not clear if the variable is supposed
to be nilable or if this is a mistake.

To fix this error, either assign nil to it here:

  @age = nil

Or declare it as nilable outside at the type level:

  @age : (Int32)?

  def initialize(@name)
      ^~~~~~~~~~
```

## Implicit types due inference might result in type mismatch in some inheritance scenarios

(aka: *why cannot use a class that inherits for this?*)

```
class Animal
end

class Dog < Animal
end

class Cat < Animal
end

class CrazyVet
  @pets : Array(Animal)

  def initialize(@pets)
    # randomly add a cat
    @pets << Cat.new
  end
end

pets = [Dog.new]
CrazyVet.new pets
```
```
in ./3449-3.cr:13: instance variable '@pets' of CrazyVet must be Array(Animal), not Array(Dog)

  def initialize(@pets)
                 ^~~~~
```

(In this particular case it might be required to check if `freeze_type` is an ancestor of `invalid_type` before give the suggestion)

---

Those are the ones that might be confusing for newcomers and saw a few issues opened several times for them.

Cheers.
